File: scripts/check-deps.js
/**
 * ExperimentIQ — Dependency checker
 * Runs automatically after `npm install` via the postinstall hook.
 * Checks for Docker and prints install instructions if missing.
 * Does NOT block the install — exits 0 regardless so npm never fails.
 */

const { execSync, spawnSync } = require("child_process");
const os = require("os");

const platform = os.platform(); // darwin | linux | win32

const GREEN  = "\x1b[32m";
const YELLOW = "\x1b[33m";
const BLUE   = "\x1b[34m";
const RESET  = "\x1b[0m";

const ok   = (msg) => console.log(`${GREEN}[✓]${RESET} ${msg}`);
const warn = (msg) => console.log(`${YELLOW}[!]${RESET} ${msg}`);
const log  = (msg) => console.log(`${BLUE}[setup]${RESET} ${msg}`);

function isAvailable(cmd) {
  const result = spawnSync(cmd, ["--version"], { stdio: "ignore" });
  return result.status === 0;
}

function isDaemonRunning() {
  const result = spawnSync("docker", ["info"], { stdio: "ignore" });
  return result.status === 0;
}

function checkDocker() {
  if (!isAvailable("docker")) {
    warn("Docker not found. ExperimentIQ requires Docker to run GrowthBook.");
    warn("");
    if (platform === "darwin") {
      warn("  Install option 1 (Homebrew):  brew install --cask docker");
      warn("  Install option 2 (direct):    https://docs.docker.com/desktop/install/mac-install/");
    } else if (platform === "linux") {
      warn("  Install guide: https://docs.docker.com/engine/install/");
      warn("  Quick install: curl -fsSL https://get.docker.com | sh");
    } else if (platform === "win32") {
      warn("  Install Docker Desktop for Windows: https://docs.docker.com/desktop/install/windows-install/");
    }
    warn("");
    warn("  After installing Docker, run:  docker compose up -d  from the project root.");
    return;
  }

  ok(`Docker found: ${execSync("docker --version").toString().trim()}`);

  if (!isDaemonRunning()) {
    warn("Docker is installed but the daemon is not running.");
    if (platform === "darwin") {
      warn("  Open Docker Desktop from your Applications folder, then run: docker compose up -d");
    } else {
      warn("  Start Docker with: sudo systemctl start docker");
    }
    return;
  }

  ok("Docker daemon is running.");
}

console.log("");
log("Checking system dependencies...");
checkDocker();
console.log("");

// Always exit 0 — never block npm install
process.exit(0);

File: scripts/check_deps.py
"""
ExperimentIQ — Python dependency checker
Run before pip install:  python scripts/check_deps.py && pip install -r backend/requirements.txt

Checks: Docker, Python version, pip
"""

import shutil
import subprocess
import sys
import platform

GREEN  = "\033[32m"
YELLOW = "\033[33m"
BLUE   = "\033[34m"
RESET  = "\033[0m"

def ok(msg):   print(f"{GREEN}[✓]{RESET} {msg}")
def warn(msg): print(f"{YELLOW}[!]{RESET} {msg}")
def log(msg):  print(f"{BLUE}[setup]{RESET} {msg}")

OS = platform.system()  # Darwin | Linux | Windows


def check_python_version():
    major, minor = sys.version_info[:2]
    if major < 3 or (major == 3 and minor < 11):
        warn(f"Python {major}.{minor} detected — ExperimentIQ requires Python 3.11+.")
        if OS == "Darwin":
            warn("  Install via Homebrew:  brew install python@3.11")
        else:
            warn("  Download from:         https://www.python.org/downloads/")
        sys.exit(1)
    ok(f"Python {major}.{minor} — OK")


def daemon_running():
    try:
        result = subprocess.run(
            ["docker", "info"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        return result.returncode == 0
    except FileNotFoundError:
        return False


def check_docker():
    if shutil.which("docker") is None:
        warn("Docker not found. ExperimentIQ requires Docker to run GrowthBook.\n")
        if OS == "Darwin":
            warn("  Install option 1 (Homebrew):  brew install --cask docker")
            warn("  Install option 2 (direct):    https://docs.docker.com/desktop/install/mac-install/")
        elif OS == "Linux":
            warn("  Quick install:  curl -fsSL https://get.docker.com | sh")
            warn("  Full guide:     https://docs.docker.com/engine/install/")
        elif OS == "Windows":
            warn("  Docker Desktop: https://docs.docker.com/desktop/install/windows-install/")
        warn("\n  After installing, run:  docker compose up -d  from the project root.")
        return

    version = subprocess.check_output(["docker", "--version"]).decode().strip()
    ok(f"{version}")

    if not daemon_running():
        warn("Docker is installed but the daemon is not running.")
        if OS == "Darwin":
            warn("  Open Docker Desktop, then run:  docker compose up -d")
        else:
            warn("  Start with:  sudo systemctl start docker")
        return

    ok("Docker daemon is running.")


if __name__ == "__main__":
    print()
    log("Checking system dependencies...\n")
    check_python_version()
    check_docker()
    print()
    ok("All checks passed. Run:  pip install -r backend/requirements.txt")
    print()
